File: scripts/setup_revenuecat.dart
/// Script to programmatically set up RevenueCat configuration
/// 
/// Run this script to automatically configure:
/// - SweepFeed Pro entitlement
/// - Monthly and Yearly products
/// - Default offering
/// 
/// Usage:
/// ```bash
/// dart scripts/setup_revenuecat.dart
/// ```
/// 
/// Or from Flutter:
/// ```bash
/// flutter run -d macos scripts/setup_revenuecat.dart
/// ```
library;

import 'dart:io';

import 'package:flutter_dotenv/flutter_dotenv.dart';

// Note: This is a standalone script, so we need to import the service directly
// In a real scenario, you might want to create a CLI tool or use this
// within your Flutter app

Future<void> main() async {
  print('🚀 RevenueCat Setup Script');
  print('==========================\n');

  // Load environment variables
  try {
    await dotenv.load();
    print('✅ Loaded .env file\n');
  } catch (e) {
    print('⚠️  Could not load .env file: $e');
    print('   Make sure REVENUECAT_SECRET_API_KEY is set\n');
  }

  // Check for secret API key
  final secretKey = dotenv.env['REVENUECAT_SECRET_API_KEY'];
  if (secretKey == null || secretKey.isEmpty) {
    print('❌ REVENUECAT_SECRET_API_KEY not found in .env');
    print('\nTo get your secret API key:');
    print('1. Go to RevenueCat Dashboard');
    print('2. Settings → API Keys → Secret Keys');
    print('3. Copy the secret key');
    print('4. Add to .env: REVENUECAT_SECRET_API_KEY=your_secret_key_here\n');
    exit(1);
  }

  print('📋 Setup Options:');
  print('1. Create entitlement only (SweepFeed Pro)');
  print('2. Complete setup (entitlement + products + offering)');
  print('3. Exit\n');

  stdout.write('Select option (1-3): ');
  final choice = stdin.readLineSync();

  switch (choice) {
    case '1':
      await _setupEntitlementOnly();
      break;
    case '2':
      await _setupComplete();
      break;
    case '3':
      print('Exiting...');
      exit(0);
    default:
      print('Invalid choice');
      exit(1);
  }
}

Future<void> _setupEntitlementOnly() async {
  print('\n🔧 Setting up entitlement only...\n');

  // This would require importing the service
  // For now, provide instructions
  print('To set up the entitlement programmatically, use:');
  print('\nIn your Flutter app:');
  print('```dart');
  print('final setupService = ref.read(revenueCatSetupServiceProvider);');
  print('await setupService.setupEntitlementOnly();');
  print('```\n');

  print('Or use the RevenueCatSetupService directly in your app.');
}

Future<void> _setupComplete() async {
  print('\n🔧 Complete setup requires store product IDs.\n');

  stdout.write('Monthly product ID (Android): ');
  final monthlyAndroid = stdin.readLineSync();

  stdout.write('Yearly product ID (Android): ');
  final yearlyAndroid = stdin.readLineSync();

  stdout.write('Monthly product ID (iOS): ');
  final monthlyIOS = stdin.readLineSync();

  stdout.write('Yearly product ID (iOS): ');
  final yearlyIOS = stdin.readLineSync();

  print('\n📝 To complete setup, use in your Flutter app:');
  print('```dart');
  print('final setupService = ref.read(revenueCatSetupServiceProvider);');
  print('await setupService.setupSweepFeedPro(');
  if (monthlyAndroid?.isNotEmpty == true) {
    print("  monthlyStoreProductIdAndroid: '$monthlyAndroid',");
  }
  if (yearlyAndroid?.isNotEmpty == true) {
    print("  yearlyStoreProductIdAndroid: '$yearlyAndroid',");
  }
  if (monthlyIOS?.isNotEmpty == true) {
    print("  monthlyStoreProductIdIOS: '$monthlyIOS',");
  }
  if (yearlyIOS?.isNotEmpty == true) {
    print("  yearlyStoreProductIdIOS: '$yearlyIOS',");
  }
  print(');');
  print('```\n');
}

File: scripts/setup_revenuecat.js
const fetch = (...args) => import('node-fetch').then(({default: fetch}) => fetch(...args));

// ==========================================
// CONFIGURATION
// ==========================================

// RevenueCat API Configuration
// Get your Secret API Key from: RevenueCat Dashboard → Settings → API Keys → Secret Keys
const REVENUECAT_SECRET_API_KEY = 'YOUR_SECRET_API_KEY_HERE';

// Get your Project ID from RevenueCat Dashboard URL or API
// It's usually in the format: app_xxxxxxxxxxxxx
const REVENUECAT_PROJECT_ID = 'YOUR_PROJECT_ID_HERE';

// Subscription Configuration
// These should match your App Store Connect subscription product IDs
const SUBSCRIPTIONS = {
  monthly: {
    identifier: 'monthly',
    iosStoreProductId: 'com.sweepfeed.app.premium.monthly',
    androidStoreProductId: 'com.sweepfeed.app.premium.monthly', // Same for now, can be different
  },
  yearly: {
    identifier: 'yearly',
    iosStoreProductId: 'com.sweepfeed.app.premium.yearly',
    androidStoreProductId: 'com.sweepfeed.app.premium.yearly',
  }
};

const ENTITLEMENT_ID = 'SweepFeed Pro';
const ENTITLEMENT_DISPLAY_NAME = 'SweepFeed Pro';
const OFFERING_ID = 'default';
const OFFERING_DISPLAY_NAME = 'Default Offering';

// ==========================================
// HELPER FUNCTIONS
// ==========================================

function getHeaders() {
  if (!REVENUECAT_SECRET_API_KEY || REVENUECAT_SECRET_API_KEY === 'YOUR_SECRET_API_KEY_HERE') {
    throw new Error('REVENUECAT_SECRET_API_KEY not configured. Get it from RevenueCat Dashboard → Settings → API Keys');
  }
  
  return {
    'Authorization': `Bearer ${REVENUECAT_SECRET_API_KEY}`,
    'Content-Type': 'application/json',
    'X-Platform': 'nodejs',
  };
}

async function apiRequest(endpoint, method = 'GET', body = null) {
  const url = `https://api.revenuecat.com/v1/projects/${REVENUECAT_PROJECT_ID}${endpoint}`;
  const options = {
    method,
    headers: getHeaders(),
  };
  if (body) options.body = JSON.stringify(body);

  const response = await fetch(url, options);
  const responseText = await response.text();
  
  if (!response.ok) {
    let errorMessage = `API Error ${response.status}: ${responseText}`;
    try {
      const errorJson = JSON.parse(responseText);
      if (errorJson.message) errorMessage = errorJson.message;
      if (errorJson.errors) errorMessage = JSON.stringify(errorJson.errors);
    } catch (e) {
      // Keep original error message
    }
    throw new Error(errorMessage);
  }
  
  try {
    return responseText ? JSON.parse(responseText) : {};
  } catch (e) {
    return {};
  }
}

// ==========================================
// MAIN SETUP FUNCTIONS
// ==========================================

async function createEntitlement() {
  console.log('\n--- Creating Entitlement ---');
  try {
    const body = {
      identifier: ENTITLEMENT_ID,
      display_name: ENTITLEMENT_DISPLAY_NAME,
    };
    
    await apiRequest('/entitlements', 'POST', body);
    console.log(`✅ Entitlement "${ENTITLEMENT_ID}" created/updated`);
    return true;
  } catch (error) {
    if (error.message.includes('409') || error.message.includes('already exists')) {
      console.log(`✅ Entitlement "${ENTITLEMENT_ID}" already exists`);
      return true;
    }
    console.error(`❌ Failed to create entitlement:`, error.message);
    return false;
  }
}

async function createProduct(identifier, storeProductId, platform) {
  console.log(`\n--- Creating Product: ${identifier} (${platform}) ---`);
  try {
    const body = {
      identifier: identifier,
      store_product_id: storeProductId,
      platform: platform,
    };
    
    await apiRequest('/products', 'POST', body);
    console.log(`✅ Product "${identifier}" created for ${platform}`);
    return true;
  } catch (error) {
    if (error.message.includes('409') || error.message.includes('already exists')) {
      console.log(`✅ Product "${identifier}" already exists for ${platform}`);
      return true;
    }
    console.error(`❌ Failed to create product:`, error.message);
    return false;
  }
}

async function attachProductToEntitlement(productIdentifier, platform) {
  console.log(`\n--- Attaching Product to Entitlement ---`);
  try {
    const body = {
      product_identifier: productIdentifier,
      platform: platform,
    };
    
    await apiRequest(`/entitlements/${ENTITLEMENT_ID}/products`, 'POST', body);
    console.log(`✅ Product "${productIdentifier}" attached to "${ENTITLEMENT_ID}" (${platform})`);
    return true;
  } catch (error) {
    if (error.message.includes('409') || error.message.includes('already')) {
      console.log(`✅ Product already attached`);
      return true;
    }
    console.error(`❌ Failed to attach product:`, error.message);
    return false;
  }
}

async function createOffering() {
  console.log('\n--- Creating Offering ---');
  try {
    const body = {
      identifier: OFFERING_ID,
      display_name: OFFERING_DISPLAY_NAME,
    };
    
    await apiRequest('/offerings', 'POST', body);
    console.log(`✅ Offering "${OFFERING_ID}" created`);
    return true;
  } catch (error) {
    if (error.message.includes('409') || error.message.includes('already exists')) {
      console.log(`✅ Offering "${OFFERING_ID}" already exists`);
      return true;
    }
    console.error(`❌ Failed to create offering:`, error.message);
    return false;
  }
}

async function addPackageToOffering(packageIdentifier, productIdentifier, platform) {
  console.log(`\n--- Adding Package to Offering ---`);
  try {
    const body = {
      identifier: packageIdentifier,
      product_identifier: productIdentifier,
      platform: platform,
    };
    
    await apiRequest(`/offerings/${OFFERING_ID}/packages`, 'POST', body);
    console.log(`✅ Package "${packageIdentifier}" added to offering (${platform})`);
    return true;
  } catch (error) {
    if (error.message.includes('409') || error.message.includes('already')) {
      console.log(`✅ Package already in offering`);
      return true;
    }
    console.error(`❌ Failed to add package:`, error.message);
    return false;
  }
}

async function setDefaultOffering() {
  console.log('\n--- Setting Default Offering ---');
  try {
    const body = {
      is_default: true,
    };
    
    await apiRequest(`/offerings/${OFFERING_ID}`, 'PATCH', body);
    console.log(`✅ Offering "${OFFERING_ID}" set as default`);
    return true;
  } catch (error) {
    console.error(`❌ Failed to set default offering:`, error.message);
    return false;
  }
}

// ==========================================
// MAIN SETUP FLOW
// ==========================================

async function setupRevenueCat() {
  console.log('🚀 RevenueCat Setup Script');
  console.log('==========================\n');
  
  // Validate configuration
  if (!REVENUECAT_SECRET_API_KEY || REVENUECAT_SECRET_API_KEY === 'YOUR_SECRET_API_KEY_HERE') {
    console.error('❌ REVENUECAT_SECRET_API_KEY not configured!');
    console.log('\nTo get your Secret API Key:');
    console.log('1. Go to https://app.revenuecat.com');
    console.log('2. Navigate to Settings → API Keys → Secret Keys');
    console.log('3. Copy your secret key');
    console.log('4. Update REVENUECAT_SECRET_API_KEY in this script\n');
    process.exit(1);
  }
  
  if (!REVENUECAT_PROJECT_ID || REVENUECAT_PROJECT_ID === 'YOUR_PROJECT_ID_HERE') {
    console.error('❌ REVENUECAT_PROJECT_ID not configured!');
    console.log('\nTo get your Project ID:');
    console.log('1. Go to RevenueCat Dashboard');
    console.log('2. Check the URL - it contains your project ID');
    console.log('3. Or go to Settings → Project Settings');
    console.log('4. Update REVENUECAT_PROJECT_ID in this script\n');
    process.exit(1);
  }

  try {
    // 1. Create Entitlement
    await createEntitlement();

    // 2. Create Products for iOS
    const monthlyIOSCreated = await createProduct(
      SUBSCRIPTIONS.monthly.identifier,
      SUBSCRIPTIONS.monthly.iosStoreProductId,
      'ios'
    );
    
    const yearlyIOSCreated = await createProduct(
      SUBSCRIPTIONS.yearly.identifier,
      SUBSCRIPTIONS.yearly.iosStoreProductId,
      'ios'
    );

    // 3. Create Products for Android (if different IDs)
    if (SUBSCRIPTIONS.monthly.androidStoreProductId !== SUBSCRIPTIONS.monthly.iosStoreProductId) {
      await createProduct(
        SUBSCRIPTIONS.monthly.identifier,
        SUBSCRIPTIONS.monthly.androidStoreProductId,
        'android'
      );
    }
    
    if (SUBSCRIPTIONS.yearly.androidStoreProductId !== SUBSCRIPTIONS.yearly.iosStoreProductId) {
      await createProduct(
        SUBSCRIPTIONS.yearly.identifier,
        SUBSCRIPTIONS.yearly.androidStoreProductId,
        'android'
      );
    }

    // 4. Attach Products to Entitlement
    if (monthlyIOSCreated) {
      await attachProductToEntitlement(SUBSCRIPTIONS.monthly.identifier, 'ios');
    }
    if (yearlyIOSCreated) {
      await attachProductToEntitlement(SUBSCRIPTIONS.yearly.identifier, 'ios');
    }

    // 5. Create Offering
    await createOffering();

    // 6. Add Packages to Offering
    if (monthlyIOSCreated) {
      await addPackageToOffering(
        SUBSCRIPTIONS.monthly.identifier,
        SUBSCRIPTIONS.monthly.identifier,
        'ios'
      );
    }
    if (yearlyIOSCreated) {
      await addPackageToOffering(
        SUBSCRIPTIONS.yearly.identifier,
        SUBSCRIPTIONS.yearly.identifier,
        'ios'
      );
    }

    // 7. Set as Default Offering
    await setDefaultOffering();

    console.log('\n✨ RevenueCat setup completed successfully!');
    console.log('\n📋 Summary:');
    console.log(`   ✅ Entitlement: ${ENTITLEMENT_ID}`);
    console.log(`   ✅ Products: monthly, yearly`);
    console.log(`   ✅ Offering: ${OFFERING_ID} (set as default)`);
    console.log('\n💡 Next Steps:');
    console.log('   1. Verify in RevenueCat Dashboard that everything looks correct');
    console.log('   2. Make sure your App Store Connect subscriptions are created');
    console.log('   3. Test purchases in your app');
    
  } catch (error) {
    console.error('\n❌ Setup Failed:', error.message);
    process.exit(1);
  }
}

// Run the setup
setupRevenueCat();
